### PR TITLE
Support variable render quantum

### LINKED ISSUE #1: Showcase example: OutputBuilder::build panicks with ALSA device
Running showcase example on ALSA device, OutputBuilder::build panicks

```rust
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: BackendSpecific { err: BackendSpecificError { description: "ALSA function 'snd_pcm_hw_params_set_buffer_size' failed with error 'EINVAL: Invalid argument'" } }', src/io.rs:81:10
stack backtrace:
   0: rust_begin_unwind
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/panicking.rs:515:5
   1: core::panicking::panic_fmt
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/panicking.rs:92:14
   2: core::result::unwrap_failed
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/result.rs:1355:5
   3: core::result::Result<T,E>::unwrap
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/result.rs:1037:23
   4: web_audio_api::io::OutputBuilder::build
             at ./src/io.rs:64:22
   5: web_audio_api::context::AudioContext::new
             at ./src/context.rs:293:22
   6: showcase::main
             at ./examples/showcase.rs:7:19
   7: core::ops::function::FnOnce::call_once
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/ops/function.rs:227:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```
If i set BUFFER_SIZE to 256, the program doesn't panick anymore.

But another bug appears. An assertion fails in debug mode (underflow in release mode)

```rust
thread 'cpal_alsa_out' panicked at 'assertion failed: `(left == right)`
  left: `78`,
 right: `0`', src/graph.rs:117:9
```
I think due to [#495](https://github.com/RustAudio/cpal/issues/495) build_output_stream is not called at fixed rate on ALSA devices.

### PR DESCRIPTION
**Pull request by Jerboas86**
`Cpal::build_output_stream` callback takes its input data from a ringbuffer which is filled by the RenderThread.
A fixed render quantum is not required anymore

The FIFO data structure comes from the [dasp](https://docs.rs/dasp/0.11.0/dasp/ring_buffer/index.html) crate.

## Caveats:
the size of the ringbuffer is fixed at 2048. if the size of the default buffer is greater than 2048, the program will crash.  

## FIFO data structures benchmarking:

**from std**
Insertions/VecDequeue      
time:   [382.99 ns 383.19 ns 383.39 ns]                              
Found 8 outliers among 100 measurements (8.00%)
  1 (1.00%) low severe
  3 (3.00%) low mild
  2 (2.00%) high mild
  2 (2.00%) high severe

**from arraydequeue crate**
Insertions/ArrayDequeue   
time:   [774.75 ns 776.21 ns 778.00 ns]    
 
 **from crossbeam-arrayqueue**                            
Insertions/ArrayQueue                                                                          
time:   [1.5268 us 1.5272 us 1.5277 us]
Found 9 outliers among 100 measurements (9.00%)
  5 (5.00%) high mild
  4 (4.00%) high severe

**from dasp::ring_buffer**
Insertions/Ringbuffer                                                                        
time:   [242.34 ns 242.50 ns 242.69 ns]
Found 6 outliers among 100 measurements (6.00%)
  4 (4.00%) high mild
  2 (2.00%) high severe

**from std**
Removes/VecDequeue    
time:   [358.29 ns 358.55 ns 358.88 ns]                           
Found 11 outliers among 100 measurements (11.00%)
  1 (1.00%) low severe
  3 (3.00%) high mild
  7 (7.00%) high severe

**from arraydequeue crate**
Removes/ArrayDequeue        
time:   [988.12 ns 995.57 ns 1.0026 us]                              
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) high mild
  2 (2.00%) high severe

 **from crossbeam-arrayqueue**  
Removes/ArrayQueue                                                                            
time:   [2.3721 us 2.3736 us 2.3753 us]
Found 14 outliers among 100 measurements (14.00%)
  8 (8.00%) high mild
  6 (6.00%) high severe

**from dasp::ring_buffer**
Removes/Ringbuffer
time:   [289.54 ns 289.82 ns 290.18 ns]                                  
Found 11 outliers among 100 measurements (11.00%)
  6 (6.00%) high mild
  5 (5.00%) high severe
 
 Fix #1

### CODE REVIEW

**Comment by orottier** [comment]
Wow thanks, you've put a lot of work in this. Let me have a good look

File: src/graph.rs [hunk 1]
```diff
@@ -129,16 +128,35 @@ impl RenderThread {
...
-                for (sample, input) in output.zip(channel) {
-                    let value = Sample::from::<f32>(input);
-                    *sample = value;
+                for (idx, sample) in channel.enumerate() {
+                    channels_buffer[i + idx * self.channels] = *sample;
                 }
             }
+
+            if self.ring_buffer.len() > 1000 {
+                panic!("debugging");
```
Comment (orottier) [inline src/graph.rs:144]: May want to drop this one

File: src/graph.rs [hunk 2]
```diff
@@ -113,13 +115,10 @@ impl RenderThread {
...
-        // assert input was properly sized
-        debug_assert_eq!(buffer.len() % chunk_size, 0);
-
-        for data in buffer.chunks_exact_mut(chunk_size) {
-            // handle addition/removal of nodes/edges
-            self.handle_control_messages();
+        // handle addition/removal of nodes/edges
+        self.handle_control_messages();
 
+        if self.ring_buffer.len() < 2 * chunk_size {
```
Comment (orottier) [inline src/graph.rs:121]: What's this magical factor `2`?

File: Cargo.toml [hunk 3]
```diff
@@ -19,6 +19,7 @@ hound = "3.4.0"
 log = "0.4.14"
 realfft = "1.1.0"
 crossbeam-channel = "0.5.1"
+dasp = {version = "0.11", features=["ring_buffer"]}
```
Comment (orottier) [inline Cargo.toml:22]: the moment we start pulling in dasp, we might as build this whole library on top of dasp_graph, dash_frame, dasp_sample etc ;)

**Comment by orottier** [comment]
Pretty cool stuff, but I think we can do even better. By buffering the last rendered quantum I think we can come up with a zero-copy, no dependency version. Let me try to cook something up

File: src/graph.rs [hunk 4]
```diff
@@ -129,16 +128,35 @@ impl RenderThread {
...
+                panic!("debugging");
+            }
+            for sample in channels_buffer {
+                self.ring_buffer.push(sample);
+            }
         }
+        // copy rendered audio into output slice
+
+        for sample in buffer.iter_mut() {
+            let input = &self.ring_buffer.pop().unwrap();
```
[the remark this answers is not in the document]
Comment (Jerboas86) [reply src/graph.rs:153]: On Alsa Host, Fixed buffer size works as a upper bound limit, so this case should not happen

File: src/graph.rs (hunk 1 above)
Comment (Jerboas86) [reply src/graph.rs:144]: Sure, i forgot to remove this one

**Comment by Jerboas86** [comment]
Sure feel free, it was a quickfix to make things work on my side

**Comment by Jerboas86** [comment]
I took into account your remarks to make a cleaner PR

**Comment by Jerboas86** [comment]
In this last commit, the dasp ring_buffer is replaced with my own implementation. DASP dep is removed
the rest of the code is the same.

perfs measurements:
|       | DASP | Custom |
|-------|------|--------|
| write | 240 ns  | 70 ns     |
| read  | 290 ns  | 150 ns   |

Todos:
- [ ] add tests to ring_buffer.rs
- [ ] make more perf measurements

**Comment by orottier** [comment]
Thanks for the updates, I'll try to have a look this weekend

**Comment by orottier** [comment]
Hey @Jerboas86, I came up with a competing implementation in https://github.com/orottier/web-audio-api-rs/pull/10
Care to take a look? This avoid copying data into a ring buffer, it just tracks the previously rendered data and a cursor

**Comment by orottier** [comment]
Closing in favour of merged #10 
Thanks for initiating this feature though!

**Pull request closed without merging**